### PR TITLE
[RFR][Feature][OSF-2718]Can't see all components when permission to one is denied

### PR DESCRIPTION
**Pull request by TomBaxter**
## Purpose:
[OSF-2718](https://openscience.atlassian.net/browse/OSF-2718)
In project overview page and project file view page...
- Enable viewing components masked by unauthorized parent.
- Allow for breadcrumbs to to get back from masked components.
In project overview page
- Update components widget to match info in file tree widget.

## Bonus
Remove visibility of unauthorized components. "Private Component"

## Changes:
Updated tests/test_views.py
* Updated various for get_children now get_readable_descendants
* Fixed test_pointer_list_read_contributor_cannot_remove_public_component_entry now actually public
* Added def test_readable_descendants_masked_by_permissions

Updated website/project/model.py
* Added def find_readable_antecedent to find logical parent
* Added def find_readable_descendants to find logical children

Updated website/project/views/node.py "def _view_project" to find logical parent and children

Updated website/routes.py added get_readable_descendants route for use by component widget

Updated website/templates/project/project.mako to utilize new get_readable_descendants route

Updated website/util/rubeus.py "def _collect_components" to find logical children and not return unauthorized nodes

## Side effects
None

[#OSF-2718]

### CODE REVIEW

**Comment by TomBaxter** [comment]
Though this test has failed twice in Travis....
FAIL: test_merge (tests.models.test_user.TestUserMerging)
It is passing for me locally and has no connection that I can see with this PR.

File: tests/test_views.py [hunk 1]
```diff
@@ -2848,25 +2848,6 @@ def _make_pointer_only_user_can_see(self, user, project, save=False):
         node = ProjectFactory(creator=user)
         project.add_pointer(node, auth=Auth(user=user), save=save)
 
-    def test_pointer_list_write_contributor_can_remove_private_component_entry(self):
```
Comment (mfraezz) [inline tests/test_views.py]: IIRC, this test existed to prevent an edge case where a pointer was created, but the only people with permission to view it were removed from the project, leading to a state where a pointer could not be deleted from the node. I'm not sure if recent changes have obviated this test, but it might still be important.

File: tests/test_views.py [hunk 2]
```diff
@@ -2903,8 +2884,9 @@ def test_pointer_list_read_contributor_cannot_remove_private_component_entry(sel
     def test_pointer_list_read_contributor_cannot_remove_public_component_entry(self):
         url = web_url_for('view_project', pid=self.project._id)
 
-        self.project.add_pointer(ProjectFactory(creator=self.user),
-                                 auth=Auth(user=self.user))
+        linked_to = ProjectFactory(creator=self.user)
+        linked_to.update(fields={'is_public':1})
```
Comment (mfraezz) [inline tests/test_views.py]: You should be able to specify arbitrary fields when using factories, as though you were instantiating an object of that type -- these three lines would be equivalent to:

```python
self.project.add_pointer(ProjectFactory(creator=self.user, is_public=True),
                         auth=Auth(user=self.user))
```

That said, this isn't a blocker

**Comment by mfraezz** [comment]
>Though this test has failed twice in Travis....
FAIL: test_merge (tests.models.test_user.TestUserMerging)
It is passing for me locally and has no connection that I can see with this PR.

Travis doesn't test just your code, it tests the code resultant after a merge into the targeted branch. See if that test breaks after a merge/rebase, there could be breaking but non-conflicting changes upstream.

File: tests/test_views.py (hunk 2 above)
Comment (TomBaxter) [inline tests/test_views.py]: Gave it a go, but got....

```
  Traceback (most recent call last):
  File "/home/tkb/work/venv/cos/osf/osf.io/tests/test_views.py", line 2891, in test_pointer_list_read_contributor_cannot_remove_public_component_entry
    self.project.add_pointer(ProjectFactory(creator=self.user), is_public=True, auth=Auth(user=self.user))
TypeError: add_pointer() got an unexpected keyword argument 'is_public'

```

  Maybe factory needs some help.

File: tests/test_views.py (hunk 1 above)
Comment (TomBaxter) [inline tests/test_views.py]: I'm not sure where to go with this. This PR has removed most, possibly all, visibility to "Private" links/components. My understanding is that a user without at least read permissions should not know about "Private" links/components, even their existence.
  This policy may be somewhat academic, but it has a practical advantage for me in this PR as I can skip "Private" breaks in the tree and show previously hidden sub-trees. I've updated the components widget to match the tree view.
  As this PR stands, I won't be able to make this test pass. Maybe there needs to be a discussion about how we want to handle this edge case. It seems like it might be a security/user experience question. Who can delete a component/link when no one has permissions to it, and where in the interface it should happen.

**Comment by TomBaxter** [comment]
I've found an issue while testing and have changed this PR to [WIP]

**Comment by mfraezz** [comment]
>I've found an issue while testing and have changed this PR to [WIP]

Moving this `Additional Development` until that's resolved, then

**Comment by TomBaxter** [comment]
Sorry for the confusion. It turned out to be a super quick fix. Ready for further review.

**Comment by TomBaxter** [comment]
During scrum, I discussed with Brian and Nici the issue with the test I removed. test_pointer_list_write_contributor_can_remove_private_component_entry. He and Nici believe that the edge case of no one being able to delete an orphaned component was fixed in another PR.

**Comment by brianjgeiger** [comment]
So @mfraezz found a problem where if you have a pointer to a node with a child you can read, and you retain access to the child but lose access to the node that the pointer points to, then you get the child node displayed. You shouldn't do this. Check for access to the node that pointers point to and, if you don't have permissions for that node, then don't check for orphaned children. The v2 api won't return pointers that you don't have access to, but v1 is not as limited.

**Comment by TomBaxter** [comment]
@brianjgeiger @mfraezz Have updated logic to properly handle links. Also tested VOLs.

File: tests/test_views.py (hunk 1 above)
Comment (mfraezz) [inline tests/test_views.py]: Would https://github.com/CenterForOpenScience/osf.io/pull/5860/commits/c3b4a96d779e4f61ab0a5037c08bf87ec6c43794 make this pass again?

Comment (TomBaxter) [inline tests/test_views.py]: For posterity. Per a meeting with @mfraezz @brianjgeiger and I, it was determined that in the component widget, we would not admit in any way that a private link exists if the user does not have read access to the node referenced by that link.

  As long as that is the case this test can't pass. Unless there is an alternative target for the test.

Comment (mfraezz) [inline tests/test_views.py]: I spoke with product about it, and the desired functionality is to maintain current behavior in the components widget for project admins only, so that they can still remove links that no one can see. No other users should be able to see the link unless they can view the node it points to.

**Comment by TomBaxter** [comment]
Visibility to links without read access added back in. Ready for review.

File: website/project/views/node.py [hunk 3]
```diff
@@ -915,6 +915,21 @@ def get_summary(auth, node, **kwargs):
...
+def get_readable_descendants(auth, node, **kwargs):
+    descendants = []
+    for child in node.nodes:
+        if child.is_deleted:
+            continue
+        elif child.can_view(auth):
+            descendants.append(child)
+        else:
+            if not child.primary:
+                if node.has_permission(auth.user, 'read'):
```
Comment (mfraezz) [inline website/project/views/node.py]: This should be `'write'`, because you need `write` access or higher to remove a `Private link`.

Nitpick: why not do `elif not child.primary and node.has_permission(auth.user, 'read'):`?

File: website/templates/project/project.mako [hunk 4]
```diff
@@ -403,7 +403,7 @@
                 <div id="containment">
                     <div mod-meta='{
                         "tpl": "util/render_nodes.mako",
-                        "uri": "${node["api_url"]}get_children/",
```
Comment (mfraezz) [inline website/templates/project/project.mako]: Was `/get_children/` being hit anywhere else? A quick grep seems to indicate that it's not.

If not, consider removing that route, method, and any tests for it. Not a blocker

Comment (TomBaxter) [inline website/templates/project/project.mako]: Done.

File: website/project/views/node.py (hunk 3 above)
Comment (TomBaxter) [inline website/project/views/node.py]: Done.

File: website/project/views/node.py [hunk 5]
```diff
@@ -915,25 +915,25 @@ def get_summary(auth, node, **kwargs):
...
+        if request.args.get('permissions'):
+            perm = request.args['permissions'].lower().strip()
+            if perm not in child.get_permissions(auth.user):
+                continue
+        if child.is_deleted:
+            continue
+        elif child.can_view(auth):
+            descendants.append(child)
+        elif not child.primary:
+                if node.has_permission(auth.user, 'write'):
```
Comment (mfraezz) [inline website/project/views/node.py]: This block should be dedented by 4 spaces

Comment (TomBaxter) [inline website/project/views/node.py]: Done.

File: website/util/rubeus.py [hunk 6]
```diff
@@ -165,8 +165,20 @@ def to_hgrid(self):
...
-            if child.resolve()._id not in visited and not child.is_deleted and node.can_view(self.auth):
+            if child.is_deleted:
+                continue
+            elif not child.can_view(self.auth):
+                if child.primary:
+                    readable_descendants = child.find_readable_descendants(self.auth)
+                    for desc in readable_descendants:
+                        visited.append(desc.resolve()._id)
+                        rv.append(self._serialize_node(desc, visited=visited))
+                else:
```
Comment (mfraezz) [inline website/util/rubeus.py]: Minor, but I think this `else` block is superfluous. It just triggers the next iteration of this loop, but if it gets hit there's nothing else that would occur this iteration anyway.

File: website/routes.py [hunk 7]
```diff
@@ -1243,11 +1243,13 @@ def make_url_map(app):
             '/project/<pid>/get_summary/',
             '/project/<pid>/node/<nid>/get_summary/',
         ], 'get', project_views.node.get_summary, json_renderer),
-
+        # Route "get_children" is deprecated. Use get_readable_descendants.
         Rule([
             '/project/<pid>/get_children/',
             '/project/<pid>/node/<nid>/get_children/',
```
Comment (mfraezz) [inline website/routes.py]: I highly doubt that any external parties are using this route, so it can probably be removed, as we're no longer using it. Not a blocker

**Comment by mfraezz** [comment]
Other than those last two nitpicky comments, LGTM. Passing this ahead 🐎

File: website/routes.py (hunk 7 above)
Comment (brianjgeiger) [inline website/routes.py]: The comment should start `# TODO: [#OSF-6557] Route…`

File: website/util/rubeus.py (hunk 6 above)
Comment (TomBaxter) [inline website/util/rubeus.py]: Done.

File: website/routes.py (hunk 7 above)
Comment (TomBaxter) [inline website/routes.py]: Done.

**Comment by TomBaxter** [comment]
Travis failure was the usual intermittent "returns false for date more than 4 years in the future"

File: website/project/model.py [hunk 8]
```diff
@@ -1267,6 +1267,36 @@ def has_permission_on_children(self, user, permission):
...
+        while next_parent:
+            if next_parent.can_view(auth):
+                return next_parent
+            next_parent = next_parent.parent_node
+
+    def find_readable_descendants(self, auth):
+        """ Returns list of first descendant node readable by <user> in
+        each descendant branch.
+        """
+        ret = []
```
Comment (sloria) [inline website/project/model.py]: In general, it's better for methods like this to be generators rather than returning a list. Leave it up to the caller if they want a list (or some other type of collection).

```python
descendants = set(node.find_readable_descendants(auth)
# or
for node in node.find_readable_descendants(auth):  # we don't load up everything in memory at once
   #..
```

**Comment by sloria** [comment]
Pass finished. Just one minor refactor to make: https://github.com/CenterForOpenScience/osf.io/pull/5860/files#r69187381. Once that's addressed, this will be good to merge for next improvements release.
